build-system/tasks/helpers.js

```diff
@@ -453,42 +472,34 @@ async function compileUnminifiedJs(srcDir, srcFilename, destDir, options) {
     };
   }
 
-  /**
-   * Bundles an entry point with all its imports and applies babel transforms.
-   * @param {boolean} continueOnError
-   */
-  async function performBundle(continueOnError) {
-    const {banner, footer} = splitWrapper();
-    await esbuild
-      .build({
-        entryPoints: [entryPoint],
-        bundle: true,
-        sourcemap: true,
-        define: experimentDefines,
-        outfile: destFile,
-        plugins: [esbuildBabelPlugin],
-        banner,
-        footer,
-      })
-      .catch((err) => handleBundleError(err, continueOnError, destFilename));
-    finishBundle(srcFilename, destDir, destFilename, options);
-    let name = destFilename;
-    if (options.latestName) {
-      const latestMaxName = options.latestName.split('.js')[0] + '.max.js';
-      name = `${name} → ${latestMaxName}`;
-    }
-    endBuildStep('Compiled', name, startTime);
-    const target = path.basename(destFilename, path.extname(destFilename));
-    if (UNMINIFIED_TARGETS.includes(target)) {
-      await applyAmpConfig(
-        destFile,
-        /* localDev */ true,
-        /* fortesting */ options.fortesting
-      );
-    }
-  }
+  const {banner, footer} = splitWrapper();
+  const buildResult = await esbuild
+    .build({
+      entryPoints: [entryPoint],
+      bundle: true,
+      sourcemap: true,
+      define: experimentDefines,
+      outfile: destFile,
+      plugins: [esbuildBabelPlugin],
+      banner,
+      footer,
+      incremental: !!options.watch,
+    })
+    .catch((err) =>
+      handleBundleError(err, options.continueOnError, destFilename)
+    );
+  finishBundle(srcFilename, destDir, destFilename, options, startTime);
 
-  await performBundle(options.continueOnError);
+  if (options.watch) {
+    watchedExtensions.set(entryPoint, {
+      rebuild: async () => {
+        const time = Date.now();
+        await buildResult.rebuild();
+        await finishBundle(srcFilename, destDir, destFilename, options, time);
+        options?.onWatchBuild();
```

Comment:
`onWatchBuild()` takes a promise that contains the result of the build. This enables the lazy-builder to wait for a pending build.

Earlier, we were passing in the result of `performBundle()`. Now, we should create a promise for the two steps on lines 497 and 498 and pass it to `onWatchBuild()`, so that if the lazy-builder is asked to trigger a new build when a build is already in progress, it will simply wait on the in-progress build.

Reply:
Updated!

Reply:
Although, from my reading of the code, I don't think it is actually necessary. `await build.pendingBuild` is called, which I think does essentially the same?

https://github.com/ampproject/amphtml/blob/2acbd7db88491ca43b49364de4610f235a09f26b/build-system/server/lazy-build.js#L91-L101

Reply:
I think the entire `onWatchBuild` flow may be unnecessary for unminified builds, since it is very safe (and even done by `esbuild`'s server to rebuild entrypoints on each request